prototype2/common/DumpFile.h

```diff
@@ -0,0 +1,181 @@
+  auto Ret = std::unique_ptr<DumpFile>(new DumpFile(FilePath, MaxMB));
+  Ret->openRW();
+  return Ret;
+}
+
+template<typename T>
+std::unique_ptr<DumpFile<T>> DumpFile<T>::open(const boost::filesystem::path& FilePath) {
+  auto Ret = std::unique_ptr<DumpFile>(new DumpFile(FilePath, 0));
+  Ret->openR();
+  return Ret;
+}
+
+template<typename T>
+boost::filesystem::path DumpFile<T>::get_full_path() const {
+  auto Ret = PathBase;
+  if (SequenceNumber > 0)
+    Ret += "_" + std::to_string(SequenceNumber);
+  Ret += ".h5";
+  return Ret;
+}
+
+template<typename T>
+void DumpFile<T>::openRW() {
+  using namespace hdf5;
+
+  File = file::create(get_full_path(), file::AccessFlags::TRUNCATE);
+
+  property::DatasetCreationList dcpl;
+  dcpl.layout(property::DatasetLayout::CHUNKED);
+  dcpl.chunk({ChunkSize});
+
+  DataSet = File.root().create_dataset(T::DatasetName(), DataType,
+                                       dataspace::Simple({0}, {dataspace::Simple::UNLIMITED}), dcpl);
+  DataSet.attributes.create_from("format_version", T::FormatVersion());
+  DataSet.attributes.create_from("EFU_version", efu_version());
+  DataSet.attributes.create_from("EFU_build_string", efu_buildstr());
+}
+
+template<typename T>
+void DumpFile<T>::openR() {
+  using namespace hdf5;
+
+  File = file::open(get_full_path(), file::AccessFlags::READONLY);
+  DataSet = File.root().get_dataset(T::DatasetName());
+  // if not initial version, expect it to be well formed
+  if (T::FormatVersion() > 0) {
+    uint16_t ver {0};
+    DataSet.attributes["format_version"].read(ver);
+    if (ver != T::FormatVersion()) {
+      throw std::runtime_error("DumpFile version mismatch");
+    }
+  }
+}
+
+template<typename T>
+size_t DumpFile<T>::count() const {
+  return hdf5::dataspace::Simple(DataSet.dataspace()).current_dimensions().at(0);
+}
+
+template<typename T>
+void DumpFile<T>::write() {
+  Slab.offset(0, count());
+  Slab.block(0, Data.size());
+  DataSet.extent({count() + Data.size()});
+  DataSet.write(Data, Slab);
+
+  if (MaxSize && (count() >= MaxSize)) {
+    SequenceNumber++;
+    openRW();
+  }
+}
+
+template<typename T>
+void DumpFile<T>::readAt(size_t Index, size_t Count) {
+  Slab.offset(0, Index);
+  Slab.block(0, Count);
+  Data.resize(Count);
+  DataSet.read(Data, Slab);
+}
+
+template<typename T>
+void DumpFile<T>::read(const boost::filesystem::path &FilePath, std::vector<T> &ExternalData) {
+  auto TempFile = DumpFile::open(FilePath);
+  TempFile->readAt(0, TempFile->count());
+  ExternalData = std::move(TempFile->Data);
+}
+
+template<typename T>
+void DumpFile<T>::push(const T& Hit) {
+  Data.push_back(Hit);
+  if (Data.size() >= ChunkSize) {
+    write();
+    Data.clear();
+  }
+}
+
+template<typename T>
+void DumpFile<T>::push(const std::vector<T>& Hits) {
+  Data.insert(Data.end(), Hits.begin(), Hits.end());
+  if (Data.size() >= ChunkSize) {
+    write();
+    Data.clear();
+  }
+}
```

Comment:
I don't know anything about boost filesystem or hdf5 libraries. But here's a question
do we need a destructor to close the file in a 'proper' way?

Reply:
No, we do not. RAII takes care of this -- design decisions of h5cpp are paying off

Reply:
Fine with me, but it certainly also hides the behaviour ;-)

Reply:
Just to be clear, `Data` is not the file, that's just the buffer array. The file closes when the class destructs. All resources are released in the correct order upon destruction, so even in case of thrown exception everything is cleaned up. That's the expected way.